keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+In order to discover the components and dependencies of an Application, UI designers can do the following.
+
+1. [Select the Application's components](#selecting-components). Note that the Application's components may also 
+aggregate sub-components. For instance, a Deployment aggregates ReplicaSets, and those ReplicaSets aggregate Pods. 
+UI designers must decide for themselves the best way to view the hierarchical structure of an application. As UI 
+designers are already faced with this challenge for objects like Deployment, the addition of the Application object 
+should not present any additional, undue burden.
+1. [Retrieve the Application's dependencies](#retrieving-dependencies). Again, designers will have to decide 
+how they wish to deal with the hierarchical structure of an aggregated application.
+
+#### Discovering an Object's Application 
+1. For objects that have a `kubernetes.io/application` annotation UI designers can 
+[retrieve the Application for the object](#retrieving-applications) as described above.
+
+#### Application Status
+
+1. For Applications that provide a [health check](#health-checks), UIs may use the `Status.Ready` condition to 
+surface the readiness of the application to users. If an Application has no health check, the `Status.Ready` field 
+will always be `UNKNOWN`.
+1. UIs may surface the installation status of an Application's components and dependencies by examining the set of 
+installed components in the Application's `Status.Installed` field.
+
+### Risks and Mitigations
+
+1. In order for the Application API to be useful, we must have broad adoption. In order to mitigate the risk of 
+a lack of adoption, we will gather consensus from developers in SIG Apps prior to promoting the API to Beta.
+1. As discussed in [Drawbacks](#drawbacks), our approach to the implementation of the Application Controller may 
+increase the CPU and memory footprint of the Controller Manager process. We will mitigate this risk by mirroring the 
+efficient implementation of the Garbage Collector.
+
+## Graduation Criteria
+
+1. The graduation criteria from Alpha to Beta is simply the consensus of opinion among SIG Apps and SIG CLI that the 
+feature is ready to be consumed by tool and UI developers. We expect developers to prototype against the API at alpha 
+level stability, but, as it will be disabled by default, we do not expect to be able to gather feedback from end users.
+2. The graduation criteria from Beta to GA is adoption and interoperability. Before graduating the feature to GA we 
+would like to see broad adoption by tools (e.g. Helm, Bitnami installer) and UIs (Kubernetes OSS UI, GKE, OpenShift, 
+Azure).
+
+## Implementation History
+
+### Planned
+
+- Kubernetes 1.10 - Initial implementation of the Application Kind and Application Controller in the apps/v1alpha 
+Group Version.
+- Kubernetes 1.11 - Promotion of the Application Kind and Application Controller to the apps/v1beta1 Group Version. The 
```

Comment:
This is a notional plan, it may be ambitious, and it certainly contains risk. If we can't hit beta in 1.11 we can update accordingly. As discussed above, I agree that we should promote it to beta only after developers have gotten a chance to test it in alpha clusters, and after they feel that they are ready to produce software to interact with it so that they can gather feedback from their own users.

Reply:
It's worth noting that there is a delay in real world usage after a release. Many folks aren't even using 1.9, yet. Tools take time to make the switch when a new version of k8s comes out, too.

Reply:
Yes. The reality is that most production clusters won't upgrade to 1.9.x until 1.10.0 is shipped.